File: Ruby/P4/Irrgarten/monster.rb
#encoding:utf-8

=begin

    Autor: Isabel Morro Tabares 

    Funcionalidades del monstruo

=end

require_relative 'dice'
require_relative 'labyrinth_character'

module Irrgarten
    class Monster < LabyrinthCharacter

    private 

        @@INITIAL_HEALTH = 5

    public 

        # Constructor 
        def initialize (name, intelligence, strength)
            super
            set_health(@@INITIAL_HEALTH)
        end

        # Métodos públicos

        def attack
            Dice.intensity(@strength)
        end

        def defend (received_attack)
            is_dead = dead()
            if (!is_dead)
                defensive_energy = Dice.intensity(@intelligence)
                if (defensive_energy < received_attack)
                    got_wounded()
                    is_dead = dead()
                end
            end
            is_dead
        end

        def to_s
            super + "]\n"
        end

    end
end

File: Java/P5/P5/irrgarten/irrgarten/src/irrgarten/Monster.java

package irrgarten;

/**
 * @brief   Funcionalidades de los monstruos
 * @author  Isabel Morro Tabares
 */

public class Monster extends LabyrinthCharacter {
    
    // Variables privadas
    
    private static final int INITIAL_HEALTH = 5;
    
    // Constructor
    
    public Monster (String name, float intelligence, float strength){
        super(name, intelligence, strength, INITIAL_HEALTH);
        setPos(-1, -1);
    }
 
    // Métodos públicos
       
    /**
     * @brief Genera el resultado del combate
     * @return La fuerza del monstruo
     */
    @Override
    public float attack(){
        return Dice.intensity(getStrength());
    }
    
    @Override
    public boolean defend (float receivedAttack){
        boolean isDead = dead();
        if (!isDead){
            float defensiveEnergy = Dice.intensity(getIntelligence());
            if (defensiveEnergy < receivedAttack){
                gotWounded();
                isDead = dead();
            }
        }
        return isDead;
    }
    
    @Override
    public String toString(){
        return super.toString() + "]";
    }
}
